cpp/tests/streaming/test_memory_reserve_or_wait.cpp

```diff
@@ -0,0 +1,224 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+#include <atomic>
+#include <thread>
+
+#include <gtest/gtest.h>
+
+#include <rapidsmpf/streaming/core/context.hpp>
+#include <rapidsmpf/streaming/core/memory_reserve_or_wait.hpp>
+#include <rapidsmpf/utils.hpp>
+
+#include "base_streaming_fixture.hpp"
+
+using namespace rapidsmpf;
+using namespace rapidsmpf::streaming;
+
+class StreamingMemoryReserveOrWait : public BaseStreamingFixture {
+  public:
+    // Create a buffer resource that can be controlled by set_mem_avail()
+    // and get_mem_avail().
+    void SetUp() override {
+        auto dev_mem_available = [this]() -> std::int64_t {
+            return mem_avail_.load(std::memory_order_acquire);
+        };
+        SetUpWithThreads(1, {{MemoryType::DEVICE, dev_mem_available}});
```

Comment:
I think this test needs to be parameterized on multiple threads in the pool. I've uncovered several deadlocks that way